### PR TITLE
Create DecompilerPrinter and ReplPrinter

### PR DESCRIPTION
**Pull request by nicolasstucki**
Enable flag by default for the decompiler

### CODE REVIEW

**Comment by nicolasstucki** [comment]
From this 
screen shot 2017-12-22 at 09 41 55
to this
screen shot 2017-12-22 at 09 41 39 1

**Comment by allanrenucci** [comment]
Not sure it is the correct solution. It prints invalid trees after lambda lift:
```scala
class Test {
  def test = {
    val y = 2
    (x: Int) => x + y
  }
}
```
```scala
package <empty> {
  @scala.annotation.internal.SourceFile("tests/allan/Test.scala") class Test
     extends
   Object { 
    def <init>(): Unit = 
      {
        super()
        ()
      }
    def test(): Function1 = 
      {
        val y: Int = 2
        {
          
        }
      }
    (y$1: Int, x: Int) => 
      {
        x.+(y$1)
      }
  }
}
```

**Comment by nicolasstucki** [comment]
This option is not supposed to be used after lambda lift. In fact it is only intended for pickling level of tree abstractions. I could make it a Y setting.

**Comment by allanrenucci** [comment]
Can't we have a solution that works across all phases? The way we currently print closures doesn't feel natural to me:
```scala
    def test(): Function1 = 
      {
        val y: Int = 2
        {
          closure(y | <empty>.this.test$$anonfun$1:JFunction1$mcII$sp)
        }
      }
    private <static> def test$$anonfun$1(y$1: Int, x: Int): Int = 
      {
        x.+(y$1)
      }
```
It would be great to print them like scalac does (at least under an option).

**Comment by nicolasstucki** [comment]
It should print the normal output if we are in a transformation phase. No normal user should ever see the trees in those.

**Comment by allanrenucci** [comment]
Don't you don't think this:
```scala
    def test(): Function1 = 
      {
        val y: Int = 2
        {
          closure(y | <empty>.this.test$$anonfun$1:JFunction1$mcII$sp)
        }
      }
```
would be better printed as:
```scala
    def test(): Function1 = 
      {
        val y: Int = 2
        {
          (x: Int) => <empty>.this.test$$anonfun$1(y: Int, x: Int)
        }
      }
```
It is still the same tree, just a different way to print it

**Comment by nicolasstucki** [comment]
@allanrenucci changing printers in transforms is outside of the scope of this PR

**Comment by allanrenucci** [comment]
If the scope of this PR is only the decompiler then I don't think it is worth the added complexity. I would rather create a special printer for the decompiler (a subclass of `RefinedPrinter`) and not introduce a new setting.

**Comment by nicolasstucki** [comment]
There would be a lot of duplication if we do not put it in the `RefinedPrinter`. And we would have to maintain any change in two parallel printers.

**Comment by smarter** [comment]
Note that we also have a https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/printing/UserFacingPrinter.scala (which I'd like to get rid of if possible), so calling this option -Yprint-user is confusing.

**Comment by nicolasstucki** [comment]
We should probably mode that logic to `UserFacingPrinter` as well.

What about `-Yprint-decompile`

**Comment by nicolasstucki** [comment]
@smarter all branches of the code in `UserFacingPrinter` were covered by tests.

File: compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala [hunk 1]
```diff
@@ -27,6 +28,10 @@ class RefinedPrinter(_ctx: Context) extends PlainPrinter(_ctx) {
   private[this] var myCtx: Context = _ctx
   private[this] var printPos = ctx.settings.YprintPos.value
   private[this] val printLines = ctx.settings.printLines.value
+  private[this] val YprintDecompile =
+    ctx.settings.YprintDecompile.value && !ctx.phases.dropWhile(!_.isInstanceOf[FirstTransform]).contains(ctx.phase)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala]: Add a comment to explain the latter condition?

**Comment by allanrenucci** [comment]
> I've a general concern to make RefinedPrinter use-scenario-aware: repl, decompile are both specific usage of RefinedPrinter.

I tend to agree with @liufengyun here. Should we discuss this PR during the next Dotty meeting?

**Comment by nicolasstucki** [comment]
First 3 commits from #3701

File: compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala [hunk 2]
```diff
@@ -14,7 +14,8 @@ import TypeApplications._
 import Decorators._
 import config.Config
 import util.Positions._
-import transform.SymUtils._
+import dotty.tools.dotc.transform.SymUtils._
+import dotty.tools.dotc.transform.FirstTransform
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala]: No, confused the IDE

**Comment by nicolasstucki** [comment]
@liufengyun the last commit is ready for review. The first 3 are based on another PR.

File: compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala [hunk 3]
```diff
@@ -64,7 +78,9 @@ class RefinedPrinter(_ctx: Context) extends PlainPrinter(_ctx) {
   protected val PrintableFlags = (SourceModifierFlags | Label | Module | Local).toCommonFlags
 
   override def nameString(name: Name): String =
-    if (ctx.settings.YdebugNames.value) name.debugString else name.toString
+    if (name.isReplAssignName) name.decode.toString.takeWhile(_ != '$')
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala]: Override `nameString` in `ReplPrinter`?

File: compiler/src/dotty/tools/dotc/decompiler/DecompilationPrinter.scala [hunk 4]
```diff
@@ -14,23 +18,41 @@ class DecompilationPrinter extends Phase {
   override def phaseName: String = "decompilationPrinter"
 
   override def run(implicit ctx: Context): Unit = {
-    val unit = ctx.compilationUnit
+    val outputDir = ctx.settings.outputDir.value
+    if (outputDir == ".") printToOutput(System.out)
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/decompiler/DecompilationPrinter.scala]: Yes, as it is the default it is practical.

File: compiler/src/dotty/tools/dotc/decompiler/DecompilationPrinter.scala [hunk 5]
```diff
@@ -14,23 +18,41 @@ class DecompilationPrinter extends Phase {
...
+    else {
+      var os: OutputStream = null
+      var ps: PrintStream = null
+      try {
+        os = File(outputDir + ".decompiled").outputStream()
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/decompiler/DecompilationPrinter.scala]: Oh, I use the `outputDir`  as a file. Just like we do with JARs, where we write `-d Foo.jar`. We can change the format later if needed.

File: compiler/src/dotty/tools/dotc/printing/DecompilerPrinter.scala [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import dotty.tools.dotc.printing.Texts._
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.core.Symbols._
+
+import scala.language.implicitConversions
+
+class DecompilerPrinter(_ctx: Context) extends RefinedPrinter(_ctx) {
+
+  override protected def modText(mods: untpd.Modifiers, kw: String): Text = { // DD
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/printing/DecompilerPrinter.scala]: I will go for the second. Documenting all changes might end in the large and unmaintainable list of features when we fix more stuff. I would rely more on the commit history and the tests added for that change.

File: compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala]: Yes, missed it.

File: compiler/src/dotty/tools/dotc/decompiler/DecompilationPrinter.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/decompiler/DecompilationPrinter.scala]: This mode is mostly intended for testing

**Pull request merged**